pkg/gardenlet/controller/shoot/shoot_care_control.go

```diff
@@ -143,10 +142,38 @@ type CareControlInterface interface {
 // implements the documented semantics for caring for Shoots. You should use an instance returned from NewDefaultCareControl()
 // for any scenario other than testing.
 func NewDefaultCareControl(clientMap clientmap.ClientMap, k8sGardenCoreInformers gardencoreinformers.Interface, secrets map[string]*corev1.Secret, imageVector imagevector.ImageVector, identity *gardencorev1beta1.Gardener, gardenClusterIdentity string, config *config.GardenletConfiguration) CareControlInterface {
-	return &defaultCareControl{clientMap, k8sGardenCoreInformers, secrets, imageVector, identity, gardenClusterIdentity, config}
+	return &defaultCareControl{
+		NewOperation,
+		NewHealthCheck,
+		NewConstraintCheck,
+		NewGarbageCollector,
+		clientMap,
+		k8sGardenCoreInformers,
+		secrets,
+		imageVector,
+		identity,
+		gardenClusterIdentity,
+		config,
+	}
 }
 
+var (
+	// NewOperation is used to create a new `operation.Operation` instance.
+	NewOperation = defaultNewOperationFunc
+	// NewHealthCheck is used to create a new Health check instance.
+	NewHealthCheck = defaultNewHealthCheck
+	// NewConstraintCheck is used to create a new Constraint check instance.
+	NewConstraintCheck = defaultNewConstraintCheck
+	// NewGarbageCollector is used to create a new Constraint check instance.
+	NewGarbageCollector = defaultNewGarbageCollector
+)
+
 type defaultCareControl struct {
+	newOperation        NewOperationFunc
+	newHealthCheck      NewHealthCheckFunc
+	newConstraintCheck  NewConstraintCheckFunc
+	NewGarbageCollector NewGarbageCollectorFunc
+
```

Comment:
```suggestion
```

Is this on purpose or type that `NewGarbageCollector` starts with a `N` while all other start with `n`?
Generally, what's the reason that those are attributes on the `defaultCareControl` structure? You have exported variables for those functions already in lines 160-169, I guess that's enough for overriding them in the test? Something like:
https://github.com/gardener/gardener/blob/bc3c8c60e9a84d85816f5135b47815cf51cecae6/pkg/operation/botanist/backupentry/backupentry_test.go#L121

Reply:
> Is this on purpose or type that NewGarbageCollector starts with a N while all other start with n?
Thanks, it wasn't on purpose. I'll change it to `newGarbageCollector `.

> Generally, what's the reason that those are attributes on the defaultCareControl structure?
The reason is that I only wanted to make these things injectable during object construction, i.e. once the `defaultCareControl` was created the behavior will never change. I can also make it using the exported vars every time if you think that's the cleaner way?

Reply:
Yeah, at least that's what we do for other packages. If there is no other reason to keep these attributes in the structure then I'd vote for removing them and only use the function variables in the package and overwrite them in the test.

Reply:
Okay, no problem :)

Reply:
Great, thanks!